Augment text extraction support to include information about the selection, links and editable elements

whsieh opened the pull request:
#### 4bb58f66236e7ff4fe1ad111373ad063c5a56b2e
<pre>
Augment text extraction support to include information about the selection, links and editable elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=269682">https://bugs.webkit.org/show_bug.cgi?id=269682</a>
<a href="https://rdar.apple.com/123201244">rdar://123201244</a>

Reviewed by Aditya Keerthi.

Extend support for text extraction to match the latest version of the corresponding system APIs. See
comments below for more details.

Test: fast/text-extraction/basic-text-extraction.html

* LayoutTests/TestExpectations:
* LayoutTests/fast/text-extraction/basic-text-extraction-expected.txt: Added.
* LayoutTests/fast/text-extraction/basic-text-extraction.html: Added.

Add a new layout test to exercise this logic (enabled on WebKit2 macOS and iOS only). This test
includes exercises all of the different item subclasses, as well as some more interesting scenarios:
(1) text, links and images nested inside of other structural DOM elements (lists, list items), and
(2) links inside of contenteditable elements.

* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestTextExtraction):
(window.UIHelper):

Add a new `UIHelper` method to dump the text extraction tree as text, for testing purposes.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:

Make several changes to the underlying model objects:

1.  For the container types, remove `Link` (now that it&apos;s represented purely as text).

2.  Also, turn `Button` into a container type and remove `InteractiveItemData`; this is because
    there is not currently a system type that represents buttons/interactive elements.

3.  Finally, revamp how `TextItemData` works:
    a.  Treat all links as ranges in text, represented by a list of `(URL, Range)` per text item.
    b.  Add an optional `selectedRange` member that represents what part of the text item is in the
        user&apos;s current selection.
    c.  Remove `EditableItemData` below, and instead add an optional `editable` member to each
        text item that contains metadata about the editing host.

To match platform expectations, we now merge adjacent text items together during traversal, with
links and selection ranges embedded inside of text items. When encountering an editable container or
host (e.g. `input`, or contenteditable `div`) we now *only* extract text items underneath that
subtree, using the aforementioned coalescing mechanism to ensure that all text items in the subtree
get appended to a single item with the `editable` member. We apply a similar treatment to links as
well, extracting only text underneath them.

(WebCore::TextExtraction::collectText):

Refactor this `collectText` method, such that it now returns a `HashMap` of all text nodes with
visible text in the document, as well as ranges in the visible text that are selected by the user.
To achieve this, if there is a ranged selection in the document, we&apos;ll divide the text iteration
into three phases: (1) collecting text before the start of the selection, (2) collecting text inside
of the selection, and (3) collecting text after the end of the selection. These three results are
then stitched together to produce the above `HashMap` with per-text-node selection ranges.

(WebCore::TextExtraction::TraversalContext::shouldIncludeNodeWithRect const):

Add a `TraversalContext` struct to help encapsulate tree state while traversing the DOM to build up
text extraction results. This now includes `onlyCollectTextAndLinksCount`, which is set when
traversing into links or editable containers.

(WebCore::TextExtraction::canMerge):
(WebCore::TextExtraction::merge):

Add helper methods to merge text nodes into adjacent nodes.

(WebCore::TextExtraction::labelText):
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::extractRecursive):
(WebCore::TextExtraction::pruneRedundantItemsRecursive):

To avoid emitting too many spurious white space nodes, add one final pass through the extraction
results before returning it, to prune all non-editable text nodes that only contain whitespace.

(WebCore::TextExtraction::extractItem):
(WebCore::TextExtraction::shouldIncludeChildren): Deleted.
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:

See above for more details.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm:
(WebKit::containerType):
(WebKit::createWKTextItem):
(WebKit::createItemWithChildren):
(WebKit::createItemRecursive):
(WebKit::createItemIgnoringChildren): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:

Update the Swift and ObjC wrappers to match the changes in `TextExtractionTypes.h`.

* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
(WKTextExtractionItem.children): Deleted.
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptContext.cpp:
(UIScriptContext::asyncTaskComplete):

Add an optional `std::initializer_list&lt;JSValueRef&gt;` argument here, so that we can pass JS values as
arguments back to test runner callbacks.

* Tools/TestRunnerShared/UIScriptContext/UIScriptContext.h:
(WTR::UIScriptContext::asyncTaskComplete):
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::requestTextExtraction):
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:

Add `WebKit/UIProcess/Cocoa` to the header search paths, so that we can include
`WKTextExtractionItems.h` in API tests.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::textExtractionEnabled const):

Add a test option to enable the `_textExtractionEnabled` SPI preference.

* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h: Copied from Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h.
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.mm: Added.
(WTR::description):
(WTR::buildDescriptionIgnoringChildren):
(WTR::buildRecursiveDescription):
(WTR::recursiveDescription):

Add support for the new text extraction layout test; see above.

Canonical link: <a href="https://commits.webkit.org/275013@main">https://commits.webkit.org/275013@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c755760d4cab161c5ecfe20024f1030555ff929

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19653 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43018 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43193 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36729 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42947 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22619 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16984 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33715 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41214 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35037 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14304 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14382 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36014 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44468 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36853 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40072 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15455 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12671 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38407 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17074 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/35296 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9103 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17125 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16718 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/text-extraction/TextExtractionTypes.h:
```diff
@@ -25,47 +25,51 @@
+};
+
 struct TextItemData {
+    Vector<std::pair<URL, CharacterRange>> links;
```
Does `std::pair` need the full class definitions? / Would forward decls not work?

whsieh replied:
Hm.... I think I do need the implementations here, so that the size of `std::pair` is known. I think I can get away with removing `URL.h`, although that's just because it's already getting it from other unified sources, so I think it's probably best to keep including that explicitly.

aprotyas replied:
Yes I would say keep it in to avoid a future unified sources tragedy

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -64,30 +81,126 @@ static inline HashMap<Ref<Text>, String> collectText(ContainerNode& container)
+                continue;
+            }
+        }
+        result.add(node.ptr(), TextAndSelectedRange { text, std::nullopt });
```
Nit: Since I saw some `{ }` elsewhere in the PR and I suppose we're trending towards that direction nowadays.

```suggestion
        result.add(node.ptr(), TextAndSelectedRange { text, { } });
```

aprotyas replied:
Though feel free to ignore if you think it harms readability here

aprotyas replied:
Ditto line 134

whsieh replied:
Good catch! Yes, I've been meaning to prefer `{ }` over `std::nullopt` going forward, but it looks like I missed a spot.

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -102,92 +215,183 @@ static inline FloatRect rootViewBounds(Node& node)
+            return false;
+
+        auto& text = std::get<TextItemData>(child.data);
+        return !text.editable && text.content.template containsOnly<isASCIIWhitespace>();
```
🤯

whsieh replied:
Yeah, it's a bit funky... 😓

It's not actually totally clear why this is necessary, but I did find that if I moved it out of the surrounding lambda, it compiled without the `template`.

aprotyas reviewed (commented):
LGTM with minor comments

whsieh committed:
Augment text extraction support to include information about the selection, links and editable elements
https://bugs.webkit.org/show_bug.cgi?id=269682
rdar://123201244

Reviewed by Aditya Keerthi.

Extend support for text extraction to match the latest version of the corresponding system APIs. See
comments below for more details.

Test: fast/text-extraction/basic-text-extraction.html

* LayoutTests/TestExpectations:
* LayoutTests/fast/text-extraction/basic-text-extraction-expected.txt: Added.
* LayoutTests/fast/text-extraction/basic-text-extraction.html: Added.

Add a new layout test to exercise this logic (enabled on WebKit2 macOS and iOS only). This test
includes exercises all of the different item subclasses, as well as some more interesting scenarios:
(1) text, links and images nested inside of other structural DOM elements (lists, list items), and
(2) links inside of contenteditable elements.

* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.requestTextExtraction):
(window.UIHelper):

Add a new `UIHelper` method to dump the text extraction tree as text, for testing purposes.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:

Make several changes to the underlying model objects:

1.  For the container types, remove `Link` (now that it's represented purely as text).

2.  Also, turn `Button` into a container type and remove `InteractiveItemData`; this is because
    there is not currently a system type that represents buttons/interactive elements.

3.  Finally, revamp how `TextItemData` works:
    a.  Treat all links as ranges in text, represented by a list of `(URL, Range)` per text item.
    b.  Add an optional `selectedRange` member that represents what part of the text item is in the
        user's current selection.
    c.  Remove `EditableItemData` below, and instead add an optional `editable` member to each
        text item that contains metadata about the editing host.

To match platform expectations, we now merge adjacent text items together during traversal, with
links and selection ranges embedded inside of text items. When encountering an editable container or
host (e.g. `input`, or contenteditable `div`) we now *only* extract text items underneath that
subtree, using the aforementioned coalescing mechanism to ensure that all text items in the subtree
get appended to a single item with the `editable` member. We apply a similar treatment to links as
well, extracting only text underneath them.

(WebCore::TextExtraction::collectText):

Refactor this `collectText` method, such that it now returns a `HashMap` of all text nodes with
visible text in the document, as well as ranges in the visible text that are selected by the user.
To achieve this, if there is a ranged selection in the document, we'll divide the text iteration
into three phases: (1) collecting text before the start of the selection, (2) collecting text inside
of the selection, and (3) collecting text after the end of the selection. These three results are
then stitched together to produce the above `HashMap` with per-text-node selection ranges.

(WebCore::TextExtraction::TraversalContext::shouldIncludeNodeWithRect const):

Add a `TraversalContext` struct to help encapsulate tree state while traversing the DOM to build up
text extraction results. This now includes `onlyCollectTextAndLinksCount`, which is set when
traversing into links or editable containers.

(WebCore::TextExtraction::canMerge):
(WebCore::TextExtraction::merge):

Add helper methods to merge text nodes into adjacent nodes.

(WebCore::TextExtraction::labelText):
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::extractRecursive):
(WebCore::TextExtraction::pruneRedundantItemsRecursive):

To avoid emitting too many spurious white space nodes, add one final pass through the extraction
results before returning it, to prune all non-editable text nodes that only contain whitespace.

(WebCore::TextExtraction::extractItem):
(WebCore::TextExtraction::shouldIncludeChildren): Deleted.
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:

See above for more details.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionItem.h:
* Source/WebKit/UIProcess/Cocoa/WKTextExtractionUtilities.mm:
(WebKit::containerType):
(WebKit::createWKTextItem):
(WebKit::createItemWithChildren):
(WebKit::createItemRecursive):
(WebKit::createItemIgnoringChildren): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:

Update the Swift and ObjC wrappers to match the changes in `TextExtractionTypes.h`.

* Source/WebKit/WebKitSwift/TextExtraction/WKTextExtractionItem.swift:
(WKTextExtractionItem.children): Deleted.
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptContext.cpp:
(UIScriptContext::asyncTaskComplete):

Add an optional `std::initializer_list<JSValueRef>` argument here, so that we can pass JS values as
arguments back to test runner callbacks.

* Tools/TestRunnerShared/UIScriptContext/UIScriptContext.h:
(WTR::UIScriptContext::asyncTaskComplete):
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::requestTextExtraction):
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:

Add `WebKit/UIProcess/Cocoa` to the header search paths, so that we can include
`WKTextExtractionItems.h` in API tests.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::textExtractionEnabled const):

Add a test option to enable the `_textExtractionEnabled` SPI preference.

* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::requestTextExtraction):
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.h: Copied from Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h.
* Tools/WebKitTestRunner/cocoa/WKTextExtractionTestingHelpers.mm: Added.
(WTR::description):
(WTR::buildDescriptionIgnoringChildren):
(WTR::buildRecursiveDescription):
(WTR::recursiveDescription):

Add support for the new text extraction layout test; see above.

Canonical link: https://commits.webkit.org/275013@main
